nixos/modules/system/activation/activatable-system.nix

```diff
@@ -52,36 +46,51 @@ in
       '';
     };
   };
-  config = {
-    system.activatableSystemBuilderCommands = ''
-      echo "$activationScript" > $out/activate
-      echo "$dryActivationScript" > $out/dry-activate
-      substituteInPlace $out/activate --subst-var-by out ''${!toplevelVar}
-      substituteInPlace $out/dry-activate --subst-var-by out ''${!toplevelVar}
-      chmod u+x $out/activate $out/dry-activate
-      unset activationScript dryActivationScript
-    '';
+  config =
+    let
+      activationScript = lib.getExe (
+        pkgs.writeShellApplication {
+          name = "activate";
+          text = config.system.activationScripts.script;
+          checkPhase = "";
+          bashOptions = [ ];
+        }
+      );
+      dryActivationScript = lib.getExe (
+        pkgs.writeShellApplication {
+          name = "dry-activate";
+          text = config.system.dryActivationScript;
+          checkPhase = "";
+          bashOptions = [ ];
+        }
+      );
+    in
+    {
+      system.activatableSystemBuilderCommands =
+        # We use sed here instead of substitute(InPlace), because the substitute
+        # functions load the content of the file into a bash variable, which fails
+        # for very large activation scripts.
+        # bash
+        ''
+          cp ${activationScript} $out/activate
+          cp ${dryActivationScript} $out/dry-activate
+          sed --in-place --expression "s|@out@|''${!toplevelVar}|g" $out/activate $out/dry-activate
```

Comment:
Is `sed` (`gnused`) available by default in this script ?

Reply:
Good question, the build succeeded, so it seems that it is, but I can make it explicit to make it more robust.

Reply:
Done